lib/widget/staggered_grid.dart

```diff
@@ -0,0 +1,141 @@
+import 'package:ensemble_ts_interpreter/invokables/invokable.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_staggered_grid_view/flutter_staggered_grid_view.dart';
+
+class EnsembleStaggeredGrid extends StatefulWidget
+    with
+        Invokable,
+        UpdatableContainer,
+        HasController<StaggeredGridController, EnsembleStaggeredGridState> {
+  static const type = 'StaggeredGrid';
+
+  EnsembleStaggeredGrid({Key? key}) : super(key: key);
+
+  final StaggeredGridController _controller = StaggeredGridController();
+  @override
+  StaggeredGridController get controller => _controller;
+
+  @override
+  State<StatefulWidget> createState() => EnsembleStaggeredGridState();
+
+  @override
+  Map<String, Function> getters() {
+    return {};
+  }
+
+  @override
+  Map<String, Function> setters() {
+    return {
+      'horizontalGap': (value) =>
+          _controller.horizontalGap = Utils.optionalDouble(value),
+      'verticalGap': (value) =>
+          _controller.verticalGap = Utils.optionalDouble(value),
+      'onItemTap': (funcDefinition) => _controller.onItemTap =
+          EnsembleAction.fromYaml(funcDefinition, initiator: this),
+    };
+  }
+
+  @override
+  Map<String, Function> methods() {
+    // TODO: implement methods
+    throw UnimplementedError();
+  }
+
+  @override
+  void initChildren({List<Widget>? children, ItemTemplate? itemTemplate}) {
+    _controller.children = children;
+    _controller.itemTemplate = itemTemplate;
+  }
+}
+
+class EnsembleStaggeredGridState extends WidgetState<EnsembleStaggeredGrid>
+    with TemplatedWidgetState {
+  List<Widget>? templatedChildren;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+
+    // evaluate item-template's initial value & listen for changes
+    if (widget._controller.itemTemplate != null) {
+      registerItemTemplate(context, widget._controller.itemTemplate!,
+          evaluateInitialValue: true, onDataChanged: (List dataList) {
+        setState(() {
+          templatedChildren = buildWidgetsFromTemplate(
+              context, dataList, widget._controller.itemTemplate!);
+        });
+      });
+    }
+  }
+
+  @override
+  Widget buildWidget(BuildContext context) {
+    final items = buildItems();
+
+    return SingleChildScrollView(
+      child: StaggeredGrid.count(
+        crossAxisCount: 4,
+        mainAxisSpacing: widget._controller.verticalGap ?? 0,
+        crossAxisSpacing: widget._controller.horizontalGap ?? 0,
+        children: List.generate(items.length, (index) {
+          final item = items[index];
+
+          return StaggeredGridTile.count(
+            crossAxisCellCount: 2,
+            mainAxisCellCount: getRandomMainAxisCount,
+            child: GestureDetector(
+              onTap: () => _onItemTapped(index),
+              child: item,
+            ),
+          );
+        }),
+      ),
+    );
+  }
+
+  int get getRandomMainAxisCount {
+    return Random().nextInt(3) + 2;
```

Comment:
this will never generate a count of 1. Intentional? Don't you want to do `nextInt(4) + 1`?

Reply:
I intentionally did like that, 1 in the main axis doesn't look good. So I started to randomize it from the 2